win32ss/gdi/ntgdi/freetype.c

```diff
@@ -206,6 +378,65 @@ BYTE FASTCALL IntCharSetFromCodePage(UINT uCodePage)
     return DEFAULT_CHARSET;
 }
 
+static BOOL
+SubstituteFontRecurse(LOGFONTW* pLogFont);
+
+static __inline VOID
+FindBestFontFromList(FONTOBJ **FontObj, ULONG *MatchPenalty,
+                     const LOGFONTW *LogFont,
+                     const PLIST_ENTRY Head);
+
+static BOOL
+MatchFontName(PSHARED_FACE SharedFace, PUNICODE_STRING Name1, FT_UShort NameID, FT_UShort LangID);
+
+static BOOL
+FontLink_GetReady(
+    _Inout_ PFONTLINK pFontLink)
+{
+    FONTOBJ *pFontObj;
+    ULONG MatchPenalty;
+    UNICODE_STRING Name1;
+    PPROCESSINFO Win32Process;
+
+    if (pFontLink->bIgnoreThis)
+        return FALSE;
+
+    if (pFontLink->SharedFace)
+        return TRUE;
+
+    MatchPenalty = 0xFFFFFFFF;
+    pFontObj = NULL;
+
+    // Search private fonts
+    Win32Process = PsGetCurrentProcessWin32Process();
+    ASSERT_FREETYPE_LOCK_HELD();
+    FindBestFontFromList(&pFontObj, &MatchPenalty, &pFontLink->LogFont,
+                         &Win32Process->PrivateFontListHead);
+
```

Comment:
What if pFontObj is not NULL by now? Should we also search the system fonts (and perhaps this will change pFontObj)?

Reply:
`pFontObj` will be the best font by judging from its penalty.

Reply:
Ah, so the FindBestFontFromList with a valid pFontObj will check whether there is a "more suitable" font in the global system font list than the one already specified in pFontObj, and if so, it would update it?

Reply:
Yes, it will check.